art_internal_impl.hpp

```diff
@@ -1060,23 +1336,96 @@ class basic_inode_4 : public basic_inode_4_parent<ArtPolicy> {
   }
   UNODB_DETAIL_RESTORE_MSVC_WARNINGS()
 
-  constexpr void delete_subtree(db &db_instance) noexcept {
+  [[nodiscard]] constexpr node_ptr get_child(
+      std::uint8_t child_index) noexcept {
+    return children[child_index].load();
+  }
+
+  // N4 - position on the first child (there is always at least one child for
+  // N4).
+  [[nodiscard]] constexpr typename basic_inode_4::iter_result begin() noexcept {
+    const auto key = keys.byte_array[0].load();
+    return {node_ptr{this, node_type::I4}, key, static_cast<uint8_t>(0)};
+  }
+
+  // N4 - position on the last child (there is always at least one child for
+  // N4).
+  [[nodiscard]] constexpr typename basic_inode_4::iter_result last() noexcept {
+    const std::uint8_t child_index = this->children_count.load() - 1;
+    const auto key = keys.byte_array[child_index].load();
+    return {node_ptr{this, node_type::I4}, key, child_index};
+  }
+
+  [[nodiscard]] constexpr typename basic_inode_4::iter_result_opt next(
+      std::uint8_t child_index) noexcept {
+    const std::uint8_t nchildren = this->children_count.load();
+    const std::uint8_t next_index = child_index + 1;  // next child index
+    if (next_index >= nchildren) return parent_class::end_result;
+    const auto key = keys.byte_array[next_index].load();
+    return {{node_ptr{this, node_type::I4}, key, next_index}};
+  }
+
+  [[nodiscard]] constexpr typename basic_inode_4::iter_result_opt prior(
+      std::uint8_t child_index) noexcept {
+    if (child_index == 0) return parent_class::end_result;
+    const std::uint8_t next_index = child_index - 1;  // next child index
+    const auto key = keys.byte_array[next_index].load();
+    return {{node_ptr{this, node_type::I4}, key, next_index}};
+  }
+
+  // N4: The keys[] is ordered for N4, so we scan the keys[] in order,
+  // returning the first value GTE the given [key_byte].
+  [[nodiscard]] constexpr typename basic_inode_4::iter_result_opt gte_key_byte(
+      std::byte key_byte) noexcept {
     const auto children_count_ = this->children_count.load();
+    for (std::uint8_t i = 0; i < children_count_; ++i) {
```

Comment:
Add a TODO to explore 1) branchless 2) SIMD implementations.

(for an example of the former see `#else` section in `add_to_nonfull`, for the latter `get_insert_pos`)

Reply:
Added

  // TODO(laurynas) explore 1) branchless 2) SIMD implementations for
  // next(), prior(), get_key_byte(), and lte_key_byte().  next() will
  // be the most frequently invoked method, so that would be the place
  // to start.  Look at each of the inode types when doing this.